statistics/handle/globalstats/global_stats_async.go

```diff
@@ -0,0 +1,468 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package globalstats
+
+import (
+	"context"
+	stderrors "errors"
+	"fmt"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/statistics/handle/storage"
+	"github.com/pingcap/tidb/statistics/handle/util"
+	"github.com/pingcap/tidb/types"
+	"github.com/tiancaiamao/gp"
+	"golang.org/x/sync/errgroup"
+)
+
+type mergeItem[T any] struct {
+	item T
+	idx  int
+}
+
+type skipItem struct {
+	histID      int64
+	partitionID int64
+}
+
+// toSQLIndex is used to convert bool to int64.
+func toSQLIndex(isIndex bool) int {
+	var index = int(0)
+	if isIndex {
+		index = 1
+	}
+	return index
+}
+
+// AsyncMergePartitionStats2GlobalStats is used to merge partition stats to global stats.
+type AsyncMergePartitionStats2GlobalStats struct {
+	is                  infoschema.InfoSchema
+	globalStats         *GlobalStats
+	pool                util.SessionPool
+	cmsketch            chan mergeItem[*statistics.CMSketch]
+	fmsketch            chan mergeItem[*statistics.FMSketch]
+	histogramAndTopn    chan mergeItem[*StatsWrapper]
+	gpool               *gp.Pool
+	allPartitionStats   map[int64]*statistics.Table
+	PartitionDefinition map[int64]model.PartitionDefinition
+	tableInfo           map[int64]*model.TableInfo
+	// key is partition id and histID
+	skipPartition             map[skipItem]struct{} //
+	getTableByPhysicalIDFn    getTableByPhysicalIDFunc
+	loadTablePartitionStatsFn loadTablePartitionStatsFunc
+	exitChan                  chan struct{}
+	globalTableInfo           *model.TableInfo
+	histIDs                   []int64
+	globalStatsNDV            []int64
+	partitionIDs              []int64
+	partitionNum              int
+	skipMissingPartitionStats bool
+}
+
+// NewAsyncMergePartitionStats2GlobalStats creates a new AsyncMergePartitionStats2GlobalStats.
+func NewAsyncMergePartitionStats2GlobalStats(
```

Comment:
Can we add more comments on it especially a diagram to show how multiple goroutines(workers) work together here?
Just like what we've done for HashAgg:
<img width="740" alt="image" src="https://github.com/pingcap/tidb/assets/7499936/bc15c1ac-257a-44e8-a26e-419124872d1a">

Reply:
Done